packages/server/movement.ts

```diff
@@ -0,0 +1,242 @@
+        }
+    }
+    unit.debug.obstacles = obstacles;
+
+    // propose steering basing on gaps in visible units
+    const d = unit.direction;
+    const nearestGapAngle = (() => {
+        obstacles.sort(([a,b], [c,d]) => a-c);
+
+        // see if current direction is blocked
+        const o = obstacles.find(o => overlap(o, [d, d]));
+        // if yes, aim for its edge
+        // TODO consider own size instead of EXTRA_TURN
+        if (o) {
+            const EXTRA_TURN = 0.1;
+            return Math.abs(o[0] - d) < Math.abs(o[1] - d)
+                    ? o[0]-EXTRA_TURN
+                    : o[1]+EXTRA_TURN;
+        } else {
+            return d;
+        }
+    })();
+    
+    let separation = {x:0, y:0};
+    for (const u of otherUnitsNearby) {
+        const MAX_LOCAL_SEPARATION_FORCE = 2;
+        const SEPARATION_OVERCOMP = 1.05; // overcompensation for separation distance
+
+        let localSeparation = V.difference(currentPos, u.position);    
+        const distance = V.magnitude(localSeparation);
+
+        // TODO real sizes or buildings which are square
+        const myOwnRadius = 1;
+        const otherUnitRadius = 1;
+
+        if (distance > (myOwnRadius + otherUnitRadius) * SEPARATION_OVERCOMP)
+            continue;
+
+        // the force gets stronger the closer it is
+        const distanceFactor = distance > 0.00001 ? 0.2 / distance : MAX_LOCAL_SEPARATION_FORCE; // avoid zero distance issues
+        localSeparation.x *= distanceFactor;
+        localSeparation.y *= distanceFactor;
+
+        // clamp the local force to avoid very high impulses at close passes
+        localSeparation = V.clamp(localSeparation, MAX_LOCAL_SEPARATION_FORCE);
+        separation = V.sum(separation, localSeparation);
+
+        // push other unit apart (but only if it can move)
+        
+        //if (u.components.find(c => c.type === 'Mover')) {
+        //    u.position.x -= localSeparation.x;
+        //    u.position.y -= localSeparation.y;
+        //}
+    }
+
+    // limit maximum    
+    const MAX_SEPARATION_FORCE = 3;
+    separation = V.clamp(separation, MAX_SEPARATION_FORCE);
+
+    // push off of terrain
+    const terrainAvoidance = {x: 0, y:0};
+    {
+        const terrainNearby =
+            allTilesInfluenced
+            .map(t => gm.tiles[explode(t)] !== 0 ? t : undefined)
+            .filter(notEmpty);
+        unit.debug.terrainNearby = terrainNearby;
+
+        for (const t of terrainNearby) {
+            const diff = V.difference(currentPos, t);
+            const distance = V.magnitude(diff);
+
+            // TODO radius etc
+            if (distance > 1.5) {
+                continue;
+            }
+
+            // convert to unit vector
+            diff.x /= distance;
+            diff.y /= distance;
+            V.vecAdd(terrainAvoidance, diff);
+        }
+    }
+    unit.debug.terrainAvoidance = terrainAvoidance;
+
+    let velocity = {
+        x: Math.cos(nearestGapAngle) * 1,

```

Comment:
`* 1` 🙀  

Is it adding any value?

Reply:
Leftover from when i was figuring out the sign between -1 and +1, easier to flip; you're right, should be gone.